dynamic/message_factory.go

```diff
@@ -1,22 +1,45 @@
 package dynamic
 
 import (
+	"fmt"
 	"reflect"
 	"sync"
 
 	"github.com/golang/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/dynamicpb"
 
 	"github.com/jhump/protoreflect/desc"
 )
 
+type RegistryMessageFactory struct {
+	types *protoregistry.Types
+}
+
+func (r *RegistryMessageFactory) Resolver() *protoregistry.Types {
+	return r.types
+}
+
+func (r *RegistryMessageFactory) NewMessage(md *desc.MessageDescriptor) proto.Message {
+	fullname := protoreflect.FullName(md.GetFullyQualifiedName())
+	mt, err := r.types.FindMessageByName(fullname)
+	if err != nil {
+		panic(fmt.Sprintf("failed to create message %s", fullname))
+	}
+
+	return dynamicpb.NewMessage(mt.Descriptor())
```

Comment:
Swapping out the old message factory for this one in `dynamic.Message.unmarshal` was what finally got the unknown options test passing